In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/digit-recognizer/sample_submission.csv
/kaggle/input/digit-recognizer/train.csv
/kaggle/input/digit-recognizer/test.csv


In [2]:
train = pd.read_csv("../input/digit-recognizer/train.csv")
train

label  pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  \
0          1       0       0       0       0       0       0       0       0   
1          0       0       0       0       0       0       0       0       0   
2          1       0       0       0       0       0       0       0       0   
3          4       0       0       0       0       0       0       0       0   
4          0       0       0       0       0       0       0       0       0   
...      ...     ...     ...     ...     ...     ...     ...     ...     ...   
41995      0       0       0       0       0       0       0       0       0   
41996      1       0       0       0       0       0       0       0       0   
41997      7       0       0       0       0       0       0       0       0   
41998      6       0       0       0       0       0       0       0       0   
41999      9       0       0       0       0       0       0       0       0   

       pixel8  ...  pixel774  pixel775  pixel776  pixel777  pixel778  \
0           0  ...         0         0         0         0         0   
1           0  ...         0         0         0         0         0   
2           0  ...         0         0         0         0         0   
3           0  ...         0         0         0         0         0   
4           0  ...         0         0         0         0         0   
...       ...  ...       ...       ...       ...       ...       ...   
41995       0  ...         0         0         0         0         0   
41996       0  ...         0         0         0         0         0   
41997       0  ...         0         0         0         0         0   
41998       0  ...         0         0         0         0         0   
41999       0  ...         0         0         0         0         0   

       pixel779  pixel780  pixel781  pixel782  pixel783  
0             0         0         0         0         0  
1             0         0         0         0         0  
2             0         0         0         0         0  
3             0         0         0         0         0  
4             0         0         0         0         0  
...         ...       ...       ...       ...       ...  
41995         0         0         0         0         0  
41996         0         0         0         0         0  
41997         0         0         0         0         0  
41998         0         0         0         0         0  
41999         0         0         0         0         0  

[42000 rows x 785 columns]

In [3]:
X = train.iloc[:, 1:]
Y = train.iloc[:, 0]

# Splitting the dataset into training and validation dataset

In [4]:
from sklearn.model_selection import train_test_split

train_X, val_X, train_Y, val_Y = train_test_split(X, Y, test_size=0.2, shuffle=True, random_state=42)

In [5]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier

In [6]:
from sklearn.metrics import f1_score
from sklearn.model_selection import cross_val_score

## DecisionTreeClassifier

In [7]:
dtc = DecisionTreeClassifier()
dtc_scores = cross_val_score(dtc, train_X, train_Y, cv=5, scoring="f1_weighted")
print(dtc_scores, dtc_scores.mean())

[0.84659524 0.850675   0.8468098  0.85793056 0.84872916] 0.8501479509213347


## RandomForestClassifier

In [8]:
rfc = RandomForestClassifier(n_estimators=100)
rfc_scores = cross_val_score(rfc, train_X, train_Y, cv=5, scoring="f1_weighted")
print(rfc_scores, rfc_scores.mean())

[0.9611187  0.96009818 0.95920591 0.96438673 0.96291472] 0.9615448464954867


In [9]:
rfc = RandomForestClassifier(n_estimators=200)
rfc_scores = cross_val_score(rfc, train_X, train_Y, cv=5, scoring="f1_weighted")
print(rfc_scores, rfc_scores.mean())

[0.96381833 0.96114273 0.96276953 0.96558432 0.96337837] 0.9633386574395256


## KNeighborsClassifier

In [10]:
knc = KNeighborsClassifier(n_neighbors=5)
knc_scores = cross_val_score(knc, train_X, train_Y, cv=5, scoring="f1_weighted")
print(knc_scores, knc_scores.mean())

[0.96419839 0.96262914 0.96252366 0.96377474 0.9676175 ] 0.964148686475836


In [11]:
knc = KNeighborsClassifier(n_neighbors=10)
knc_scores = cross_val_score(knc, train_X, train_Y, cv=5, scoring="f1_weighted")
print(knc_scores, knc_scores.mean())

[0.96035584 0.96140865 0.95849291 0.95884489 0.96253622] 0.960327703349739


In [12]:
knc = KNeighborsClassifier(n_neighbors=6)
knc_scores = cross_val_score(knc, train_X, train_Y, cv=5, scoring="f1_weighted")
print(knc_scores, knc_scores.mean())

[0.9616734  0.96124788 0.96235207 0.96271909 0.96492591] 0.9625836706918978


In [13]:
knc = KNeighborsClassifier(n_neighbors=4)
knc_scores = cross_val_score(knc, train_X, train_Y, cv=5, scoring="f1_weighted")
print(knc_scores, knc_scores.mean())

[0.96193725 0.96047694 0.96175315 0.96256221 0.966415  ] 0.9626289065395406


## XGBoost

In [14]:
xgb = XGBClassifier(n_estimators=2, objective="multi:softmax")
xgb_scores = cross_val_score(xgb, train_X, train_Y, cv=5, scoring="f1_weighted")
print(xgb_scores, xgb_scores.mean())

[0.87977752 0.87730107 0.88130602 0.88486722 0.89134047] 0.8829184599174592


In [15]:
xgb = XGBClassifier(n_estimators=10, objective="multi:softmax")
xgb_scores = cross_val_score(xgb, train_X, train_Y, cv=5, scoring="f1_weighted")
print(xgb_scores, xgb_scores.mean())

[0.93444508 0.93471298 0.93281011 0.93878863 0.93535571] 0.9352225019921276


In [16]:
xgb = XGBClassifier(n_estimators=15, objective="multi:softmax")
xgb_scores = cross_val_score(xgb, train_X, train_Y, cv=5, scoring="f1_weighted")
print(xgb_scores, xgb_scores.mean())

[0.94537233 0.94316901 0.94199317 0.95010495 0.94385938] 0.9448997677397456


In [17]:
xgb = XGBClassifier(n_estimators=20, objective="multi:softmax")
xgb_scores = cross_val_score(xgb, train_X, train_Y, cv=5, scoring="f1_weighted")
print(xgb_scores, xgb_scores.mean())

[0.95189712 0.94927878 0.94882589 0.95442608 0.95130867] 0.9511473069212846


In [18]:
xgb = XGBClassifier(n_estimators=50, objective="multi:softmax", learning_rate=1)
xgb_scores = cross_val_score(xgb, train_X, train_Y, cv=5, scoring="f1_weighted")
print(xgb_scores, xgb_scores.mean())

[0.96026752 0.95939414 0.96113513 0.96188771 0.96264716] 0.9610663328808027


## GaussianNB

In [19]:
gnb = GaussianNB()
gnb_scores = cross_val_score(gnb, train_X, train_Y, cv=5, scoring="f1_weighted")
print(gnb_scores, gnb_scores.mean())

[0.50342315 0.53214448 0.53646118 0.52418215 0.52664858] 0.5245719092238923


# Best performance is shown by KNeighborsClassifier, XGBoost, and RandomForestClassifier

## KNeighborsClassifier

In [20]:
knc_model = KNeighborsClassifier()
knc_model.fit(train_X, train_Y)

KNeighborsClassifier()

In [21]:
f1_score(train_Y, knc_model.predict(train_X), average="weighted")

0.9771441018276976

In [22]:
pred_Y = knc_`model.predict(val_X)
f1 = f1_score(val_Y, pred_Y, average="weighted")
f1

0.9648451446042975

## KGBoost

In [23]:
xgb_model = XGBClassifier(n_estimators=50, objective="multi:softmax")
xgb_model.fit(train_X, train_Y)
f1_score(train_Y, xgb_model.predict(train_X), average="weighted")

1.0

In [24]:
pred_Y = xgb_model.predict(val_X)
f1 = f1_score(val_Y, pred_Y, average="weighted")
f1

0.9703512253194359

## RandomForestClassifier

In [25]:
rfc_model = RandomForestClassifier(n_estimators=200)
rfc_model.fit(train_X, train_Y)
f1_score(train_Y, rfc_model.predict(train_X), average="weighted")

1.0

In [26]:
pred_Y = rfc_model.predict(val_X)
f1 = f1_score(val_Y, pred_Y, average="weighted")
f1

0.9658188997663494

# XGBClassifier performed the best on validation dataset with an f1 score of 0.97035

## Training it on the complete training dataset + validation dataset

In [27]:
model = XGBClassifier(n_estimators=50, objective="multi:softmax")
model.fit(X, Y)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=50, n_jobs=None,
              num_parallel_tree=None, objective='multi:softmax', ...)

## Testing on the test set

In [28]:
test = pd.read_csv ("../input/digit-recognizer/test.csv")
test

pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  \
0           0       0       0       0       0       0       0       0       0   
1           0       0       0       0       0       0       0       0       0   
2           0       0       0       0       0       0       0       0       0   
3           0       0       0       0       0       0       0       0       0   
4           0       0       0       0       0       0       0       0       0   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
27995       0       0       0       0       0       0       0       0       0   
27996       0       0       0       0       0       0       0       0       0   
27997       0       0       0       0       0       0       0       0       0   
27998       0       0       0       0       0       0       0       0       0   
27999       0       0       0       0       0       0       0       0       0   

       pixel9  ...  pixel774  pixel775  pixel776  pixel777  pixel778  \
0           0  ...         0         0         0         0         0   
1           0  ...         0         0         0         0         0   
2           0  ...         0         0         0         0         0   
3           0  ...         0         0         0         0         0   
4           0  ...         0         0         0         0         0   
...       ...  ...       ...       ...       ...       ...       ...   
27995       0  ...         0         0         0         0         0   
27996       0  ...         0         0         0         0         0   
27997       0  ...         0         0         0         0         0   
27998       0  ...         0         0         0         0         0   
27999       0  ...         0         0         0         0         0   

       pixel779  pixel780  pixel781  pixel782  pixel783  
0             0         0         0         0         0  
1             0         0         0         0         0  
2             0         0         0         0         0  
3             0         0         0         0         0  
4             0         0         0         0         0  
...         ...       ...       ...       ...       ...  
27995         0         0         0         0         0  
27996         0         0         0         0         0  
27997         0         0         0         0         0  
27998         0         0         0         0         0  
27999         0         0         0         0         0  

[28000 rows x 784 columns]

In [29]:
prediction = xgb_model.predict(test)

In [30]:
pd.DataFrame(prediction).to_csv("predictions.csv")